src/core/ext/xds/xds_server_config_fetcher.cc

```diff
@@ -0,0 +1,125 @@
+// limitations under the License.
+//
+//
+
+#include <grpc/support/port_platform.h>
+
+#include "src/core/ext/xds/xds_client.h"
+#include "src/core/lib/surface/api_trace.h"
+#include "src/core/lib/surface/server.h"
+
+namespace grpc_core {
+namespace {
+
+class XdsServerConfigFetcher : public grpc_server_config_fetcher {
+ public:
+  XdsServerConfigFetcher() {
+    grpc_error* error = GRPC_ERROR_NONE;
+    xds_client_ = XdsClient::GetOrCreate(&error);
+    if (error != GRPC_ERROR_NONE) {
+      gpr_log(GPR_ERROR, "Failed to create xds client: %s",
+              grpc_error_string(error));
+    }
+  }
+
+  void StartWatch(std::string listening_address,
+                  std::unique_ptr<grpc_server_config_fetcher::WatcherInterface>
+                      watcher) override {
+    auto listener_watcher = absl::make_unique<ListenerWatcher>(watcher.get());
+    auto* listener_watcher_ptr = listener_watcher.get();
+    // TODO(yashykt): Get the resource name id from bootstrap
+    xds_client_->WatchListenerData(
+        absl::StrCat("grpc/server?udpa.resource.listening_address=",
+                     listening_address),
+        std::move(listener_watcher));
+    MutexLock lock(&mu_);
+    auto& watcher_state = watchers_[watcher.get()];
+    watcher_state.listening_address = listening_address;
+    watcher_state.watcher = std::move(watcher);
+    watcher_state.listener_watcher = listener_watcher_ptr;
+  }
+
+  void CancelWatch(
+      grpc_server_config_fetcher::WatcherInterface* watcher) override {
+    MutexLock lock(&mu_);
+    auto it = watchers_.find(watcher);
+    if (it != watchers_.end()) {
+      // Cancel the watch on the listener before erasing
+      xds_client_->CancelListenerDataWatch(it->second.listening_address,
+                                           it->second.listener_watcher,
+                                           false /* delay_unsubscription */);
+    }
+    watchers_.erase(watcher);
+  }
+
+  // Return the interested parties from the xds client so that it can be polled.
+  grpc_pollset_set* interested_parties() override {
+    return xds_client_->interested_parties();
+  }
+
+ private:
+  class ListenerWatcher : public XdsClient::ListenerWatcherInterface {
+   public:
+    explicit ListenerWatcher(
+        grpc_server_config_fetcher::WatcherInterface* server_config_watcher)
+        : server_config_watcher_(server_config_watcher) {}
+
+    void OnListenerChanged(XdsApi::LdsUpdate listener) override {
+      // TODO(yashykt): Construct channel args according to received update
+      server_config_watcher_->UpdateConfig(nullptr);
+    }
+
+    void OnError(grpc_error* error) override {
+      gpr_log(GPR_ERROR, "ListenerWatcher:%p XdsClient reports error: %s", this,
+              grpc_error_string(error));
+      GRPC_ERROR_UNREF(error);
+      // TODO(yashykt): We might want to bubble this error to the application.
+    }
+
+    void OnResourceDoesNotExist() override {
```

Comment:
I can't respond on reviewable.io due to the overbroad oauth permissions it requires.

@markdroth wrote:
> We should probably discuss how we should handle this case for the gRPC server. On the client, we report TRANSIENT_FAILURE in this case. The analogous thing here might be to stop listening, but that seems harder to implement.



The analogous thing would be to go back to the waiting state we use during start(). That would mean a graceful shutdown: stop listen and issue GOAWAYs (although for Go the listening would continue, just stop accepting). Existing RPCs can continue, because xDS configuration has already been applied to them.

Go and Java could gracefully stop the server without much trouble. But at least Java would have trouble starting it back up again for mundane reasons. I had some future API plans that would make that easier though.

Reply:
> although for Go the listening would continue, just stop accepting 
> Go and Java could gracefully stop the server without much trouble.

The ~first thing we do in `GracefulStop` in Go is stop listening:

https://github.com/grpc/grpc-go/blob/c638ab8ccda6cefc15eb9a25357fc98b018ede37/server.go#L1677-L1679

And we have no way to re-start serving currently, either.

We'd have to have a way to GracefulStop without closing the listener, and a way to re-start.  Neither of these is trivial, but would not be insurmountable either.

Reply:
Ugh. I thought you would just stop Accept()ing, since you don't "own" that listener. But I guess that is impossible as the Accept() is non-interruptable.

Preventing Close() is "easy", as you can wrap the provided Listener (yes, I know wrapping can lead to interesting details). But Accept() wouldn't return until at least one more connection arrived. I guess you could just have Accept() continue sitting there, and if a connection comes just close it.

Reply:
> I guess you could just have Accept() continue sitting there, and if a connection comes just close it.

Maybe that's the right thing to do anyway?  (Keep accepting connections, just kill them immediately.)

Reply:
I was thinking the same thing. But I thought we had a reason for not doing that already. I don't see anything in our notes though.

Reply:
I think completely stopping the server will be hard in C-core, because the xDS integration has to happen inside the server for polling reasons.  We might be able to fix that in the future, once we move to the new polling model, but I don't think we can easily do it now.

I like the idea of accepting and them immediately closing new connections.  That should work fine with pick_first.  In fact, it seems better than just not calling `accept()`, because with that approach the first few incoming connections would still have to timeout due to the kernel accept queue; if we accept the connection and then immediately close it, the clients will see the failure without waiting for a timeout.  So can we agree to go with that approach?

In the case where the resource is removed, should we also close pre-existing connections (those for which the handshake already completed before the xDS resource was removed)?

Reply:
> So can we agree to go with that approach?

Accept+close SGTM, but it isn't a complete solution. We also have to shut down existing connections. It seems like Go could do this semi-easily by making listener.Close() a noop and starting a graceful shutdown. In Java I could do a hack where a credential shuts down the connection, or we add a way to clone server configuration, or we add a large new Listener API I've been brainstorming for a few years. I'm not wild about those options. I imagine it would be awkward for C as well.

Reply:
Yeah, it is probably a little awkward for C-core too, but I think it's doable.  I suspect that we can have the `Chttp2Listener` keep a list of active connections and then send down a transport op that closes those connections.   @yashykt, what do you think?

Reply:
Actually, for Java I see a way we can fairly easily disable builder mutation after XdsServer creation, which would allow us to re-create the server. So XdsServer could just shut down the grpc server when the listener is deleted and re-create it when listener exists again. There'd be some non-trivial synchronization to implement Server interface, but it would be self-contained.

So I think there are two options on the table for Java at the moment: credential hack and server shutdown+recreation.

Reply:
@dfawley Are you okay with closing existing connections in Go?

Reply:
cc @easwars as FYI and in case I missed anything that might make this difficult.

Yes, that SGTM.  We already wrap the listener so making its Close() not close the user's listener when the grpc server is shutdown (until the xds-grpc-server is shutdown) is easy enough.

At that point, we would immediately create a new gRPC server and make the listener wrapper insta-close new connections until the LDS resource returns.